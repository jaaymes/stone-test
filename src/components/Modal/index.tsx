import * as React from 'react'

import { Alert } from '@mui/material'
import Dialog from '@mui/material/Dialog'
import DialogActions from '@mui/material/DialogActions'
import DialogContent from '@mui/material/DialogContent'
import DialogContentText from '@mui/material/DialogContentText'
import DialogTitle from '@mui/material/DialogTitle'
import Slide from '@mui/material/Slide'
import { TransitionProps } from '@mui/material/transitions'

const Transition = React.forwardRef(function Transition(
  props: TransitionProps & {
    children: React.ReactElement<any, any>
  },
  ref: React.Ref<unknown>
) {
  return <Slide direction="up" ref={ref} {...props} />
})

interface ModalProps {
  open: boolean
  onClose: () => void
  title: string
  description?: string | JSX.Element
  alert?: string
  buttons: JSX.Element | JSX.Element[]
  maxWidth?: 'xs' | 'sm' | 'md' | 'lg' | 'xl' | false
  justifyContent?: 'flex-start' | 'center' | 'flex-end' | 'space-between' | 'space-around' | 'space-evenly'
}

const Modal: React.FC<ModalProps> = ({
  open,
  onClose,
  title,
  description,
  alert,
  buttons,
  maxWidth = 'md',
  justifyContent,
}) => {
  return (
    <Dialog
      fullWidth={true}
      maxWidth={maxWidth}
      open={open}
      TransitionComponent={Transition}
      keepMounted
      onClose={onClose}
    >
      <DialogTitle>{title}</DialogTitle>
      <DialogContent sx={{ gap: 2, display: 'flex', flexDirection: 'column' }}>
        {typeof description === 'string' ? <DialogContentText>{description}</DialogContentText> : <>{description}</>}
        {alert && <Alert severity="warning">{alert}</Alert>}
      </DialogContent>
      <DialogActions sx={{ justifyContent: justifyContent || 'space-between', px: 3.5 }}>{buttons}</DialogActions>
    </Dialog>
  )
}

export default Modal
