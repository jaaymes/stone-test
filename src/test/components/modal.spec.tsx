import Modal from '@/components/Modal'

import { fireEvent, render, screen } from '@testing-library/react'

describe('Modal', () => {
  const title = 'Test Modal'
  const description = 'Test Description'
  const alert = 'Test Alert'
  const buttons = <button>Fechar</button>
  const onClose = jest.fn()

  it('renderiza o título, a descrição, o alerta e os botões corretamente', async () => {
    const { container } = render(
      <Modal open={true} onClose={onClose} title={title} description={description} alert={alert} buttons={buttons} />
    )
    const titleElement = screen.getByText(title)
    const descriptionElement = screen.getByText(description)
    const alertElement = screen.getByText(alert)
    expect(titleElement).toBeInTheDocument()
    expect(descriptionElement).toBeInTheDocument()
    expect(alertElement).toBeInTheDocument()
    const closeButton = screen.getByText('Fechar')
    expect(closeButton).toBeInTheDocument()
    fireEvent.click(closeButton)

    expect(container).toMatchSnapshot()
  })
})
