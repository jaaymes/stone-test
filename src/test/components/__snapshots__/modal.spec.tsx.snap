// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renderiza o título, a descrição, o alerta e os botões corretamente 1`] = `
<div
  aria-hidden="true"
/>
`;
